### PR TITLE
DebertaTokenizer Rework closes #10258

### LINKED ISSUE #10258: Deberta Tokenizer convert_ids_to_tokens() is not giving expected results
## Environment info

- `transformers` version: 4.3.0
- Platform: Colab
- Python version: 3.9
- PyTorch version (GPU?): No
- Tensorflow version (GPU?): No
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

## Information
I am using Deberta Tokenizer. `convert_ids_to_tokens()` of the tokenizer is not working fine. 

The problem arises when using:
* [x] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset

## To reproduce

1. Get Debrta Tokenizer 
```python
from transformers import DebertaTokenizer
deberta_tokenizer = DebertaTokenizer.from_pretrained('microsoft/deberta-base')
```
2. Encode Some Example Using Tokenizer
```python
example = "Hi I am Bhadresh. I found an issue in Deberta Tokenizer"
encoded_example = distilbert_tokenizer.encode(example)
```
3. Convert Ids to tokens:
```python
distilbert_tokenizer.convert_ids_to_tokens(encoded_example )
"""
Output: ['[CLS]', '17250', '314', '716', '16581', '324', '3447', '13', '314', '1043', '281', '2071', '287', '1024', '4835', '64', '29130', '7509', '[SEP]']
"""
```
[Colab Link For Reproducing](https://github.com/bhadreshpsavani/UnderstandingNLP/blob/master/DebertaTokenizerIssue.ipynb)
## Expected behavior

It should return some tokens like this
```
['[CLS]', 'hi', 'i', 'am', 'b', '##had', '##resh', '.', 'i', 'found', 'an', 'issue', 'in', 'de', '##bert', '##a', 'token', '##izer', '[SEP]']
```
Not just convert an integer to string like the current behavior

#### Tagging SMEs for help: 
@n1t0, @LysandreJik

### PR DESCRIPTION
**Pull request by cronoik**
# What does this PR do?

Fixes #10258

@BigBird01 Please upload these [files](https://drive.google.com/drive/folders/1gH5EMABR94iHO7SCb_AdNGCOOIdSloxh?usp=sharing) to your deberta repositories.
@huggingface: Please don't merge before @BigBird01 has uploaded the files to his repository.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

- tokenizers:  @LysandreJik, @BigBird01

### CODE REVIEW

**Comment by LysandreJik** [comment]
This is great @cronoik! I just tested it and reproduced identical results between the previous version and this one. Fantastic!

Could you add an integration tests for the DeBERTa tokenizer to ensure the implementations don't diverge? You can just copy paste the [test for ALBERT tokenizers](https://github.com/huggingface/transformers/blob/master/tests/test_tokenization_albert.py#L106) and change the values to DeBERTa.

If you don't have time, let us know and we'll take care of it.

**Comment by cronoik** [comment]
@LysandreJik 
I will give it a try.

**Comment by cronoik** [comment]
Hi,
a new test is now added for the debertatokenizer. I stuck more to the robertatokenizer test since the debertatokenizer is basically a robertatokenizer.

I can not interpret the failed check. I assume it is because the model hub is missing the new files? @LysandreJik could you please have a look?

**Comment by jeswan** [comment]
I am trying out this PR with the following:

```
tokenizer = DebertaTokenizer.from_pretrained("microsoft/deberta-base")
target_tokenized = tokenizer.tokenize("Some test text")
```
but I see the following error (`TypeError: expected str, bytes or os.PathLike object, not NoneType`):
```
    def __init__(
        self,
        vocab_file,
        merges_file,
        errors="replace",
        unk_token="<|endoftext|>",
        bos_token="<|endoftext|>",
        eos_token="<|endoftext|>",
        add_prefix_space=False,
        **kwargs
    ):
        bos_token = AddedToken(bos_token, lstrip=False, rstrip=False) if isinstance(bos_token, str) else bos_token
        eos_token = AddedToken(eos_token, lstrip=False, rstrip=False) if isinstance(eos_token, str) else eos_token
        unk_token = AddedToken(unk_token, lstrip=False, rstrip=False) if isinstance(unk_token, str) else unk_token
        super().__init__(
            errors=errors,
            unk_token=unk_token,
            bos_token=bos_token,
            eos_token=eos_token,
            add_prefix_space=add_prefix_space,
            **kwargs,
        )

>       with open(vocab_file, encoding="utf-8") as vocab_handle:
E       TypeError: expected str, bytes or os.PathLike object, not NoneType

../../../../jeswan_transformers/src/transformers/models/gpt2/tokenization_gpt2.py:179: TypeError
```
Not sure if it is related to @cronoik's error above

**Comment by cronoik** [comment]
@jeswan that will not work since the required files are not uploaded to the model hub yet. You can download them from the [link](https://drive.google.com/drive/folders/1gH5EMABR94iHO7SCb_AdNGCOOIdSloxh?usp=sharing) and load the tokenizer from local.

File: src/transformers/models/deberta/tokenization_deberta.py [hunk 1]
```diff
@@ -603,14 +152,15 @@ def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None):
         Returns:
             :obj:`List[int]`: List of `input IDs <../glossary.html#input-ids>`__ with the appropriate special tokens.
         """
-
         if token_ids_1 is None:
             return [self.cls_token_id] + token_ids_0 + [self.sep_token_id]
         cls = [self.cls_token_id]
         sep = [self.sep_token_id]
-        return cls + token_ids_0 + sep + token_ids_1 + sep
+        return cls + token_ids_0 + sep + sep + token_ids_1 + sep
```
Comment (BigBird01) [inline src/transformers/models/deberta/tokenization_deberta.py:159]: In DeBERTa, the input schema is cls + seg1 + sep + seg2 + sep

File: src/transformers/models/deberta/tokenization_deberta.py [hunk 2]
```diff
@@ -626,25 +176,21 @@ def get_special_tokens_mask(self, token_ids_0, token_ids_1=None, already_has_spe
...
-                )
-            )
+            return list(map(lambda x: 1 if x in [self.sep_token_id, self.cls_token_id] else 0, token_ids_0))
 
-        if token_ids_1 is not None:
-            return [1] + ([0] * len(token_ids_0)) + [1] + ([0] * len(token_ids_1)) + [1]
-        return [1] + ([0] * len(token_ids_0)) + [1]
+        if token_ids_1 is None:
+            return [1] + ([0] * len(token_ids_0)) + [1]
+        return [1] + ([0] * len(token_ids_0)) + [1, 1] + ([0] * len(token_ids_1)) + [1]
```
Comment (BigBird01) [inline src/transformers/models/deberta/tokenization_deberta.py:189]: The same as above comment

File: src/transformers/models/deberta/tokenization_deberta.py [hunk 3]
```diff
@@ -668,15 +214,13 @@ def create_token_type_ids_from_sequences(self, token_ids_0, token_ids_1=None):
         """
         sep = [self.sep_token_id]
         cls = [self.cls_token_id]
+
         if token_ids_1 is None:
             return len(cls + token_ids_0 + sep) * [0]
-        return len(cls + token_ids_0 + sep) * [0] + len(token_ids_1 + sep) * [1]
+        return len(cls + token_ids_0 + sep + sep + token_ids_1 + sep) * [0]
```
Comment (BigBird01) [inline src/transformers/models/deberta/tokenization_deberta.py:220]: The same

**Review by BigBird01 — CHANGES_REQUESTED** [review]
Thanks for the effort. I just checked all the changes and left one comment which is important for the model performance.

**Comment by BigBird01** [comment]
@cronoik thanks for your effort. I just uploaded the files to the model repository and left one comment to the changes.

**Comment by cronoik** [comment]
@BigBird01 Thank you for the review. I had only tested single sentences and completely ignored the sentence pairs before pushing.
@LysandreJik Can you please help me with the test error?

**Review by LysandreJik — APPROVED** [review]
I just pushed a style commit. LGTM!

**Review by sgugger — APPROVED** [review]
Thanks for the rework!

**Comment by cronoik** [comment]
@sgugger Thanks for the review. I have pushed/accepted your change requests.

**Comment by LysandreJik** [comment]
Thanks for your efforts @cronoik!

**Pull request merged**